homeassistant/helpers/restore_state.py

```diff
@@ -3,87 +3,112 @@
+
+        # Dump states when stopping hass
+        hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STOP, async_add_dump_states_job)
+
+    hass.bus.async_listen_once(
+        EVENT_HOMEASSISTANT_START, async_setup_restore_state)
+
 
-def _load_restore_cache(hass: HomeAssistant):
+async def _load_restore_cache(hass: HomeAssistant) -> None:
     """Load the restore cache to be used by other components."""
     @callback
     def remove_cache(event):
         """Remove the states cache."""
         hass.data.pop(DATA_RESTORE_CACHE, None)
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, remove_cache)
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, remove_cache)
 
-    last_run = last_recorder_run(hass)
+    store = _get_restore_state_store(hass)
 
-    if last_run is None or last_run.end is None:
-        _LOGGER.debug('Not creating cache - no suitable last run found: %s',
-                      last_run)
+    states = await store.async_load()
+    if states is None:
+        _LOGGER.debug('Not creating cache - no saved states found')
         hass.data[DATA_RESTORE_CACHE] = {}
         return
 
-    last_end_time = last_run.end - timedelta(seconds=1)
-    # Unfortunately the recorder_run model do not return offset-aware time
-    last_end_time = last_end_time.replace(tzinfo=dt_util.UTC)
-    _LOGGER.debug("Last run: %s - %s", last_run.start, last_end_time)
-
-    states = get_states(hass, last_end_time, run=last_run)
-
-    # Cache the states
     hass.data[DATA_RESTORE_CACHE] = {
-        state.entity_id: state for state in states}
+        state['entity_id']: State.from_dict(state) for state in states}
     _LOGGER.debug('Created cache with %s', list(hass.data[DATA_RESTORE_CACHE]))
 
 
+def _get_restore_state_store(hass: HomeAssistant) -> Store:
+    """Return the data store for the last session's states."""
+    if DATA_RESTORE_STORAGE not in hass.data:
+        hass.data[DATA_RESTORE_STORAGE] = hass.helpers.storage.Store(
+            STORAGE_VERSION, STORAGE_KEY, encoder=JSONEncoder)
+
+    return hass.data[DATA_RESTORE_STORAGE]
+
+
 @bind_hass
-async def async_get_last_state(hass, entity_id: str):
+async def async_get_last_state(hass: HomeAssistant, entity_id: str) -> State:
     """Restore state."""
     if DATA_RESTORE_CACHE in hass.data:
         return hass.data[DATA_RESTORE_CACHE].get(entity_id)
 
-    if _RECORDER not in hass.config.components:
-        return None
-
     if hass.state not in (CoreState.starting, CoreState.not_running):
         _LOGGER.debug("Cache for %s can only be loaded during startup, not %s",
                       entity_id, hass.state)
         return None
 
-    try:
-        with async_timeout.timeout(RECORDER_TIMEOUT, loop=hass.loop):
-            connected = await wait_connection_ready(hass)
-    except asyncio.TimeoutError:
-        return None
-
-    if not connected:
-        return None
-
     if _LOCK not in hass.data:
```

Comment:
No need to do a lock. Instead we can use the trick we used for the entity registry and share the task to create + load the store: 

https://github.com/home-assistant/home-assistant/blob/419725e1a9026afe63640426ac14ae5124d91cc1/homeassistant/helpers/entity_registry.py#L257-L270